text/chapter10.md

```diff
@@ -391,416 +437,935 @@ foreign import isUndefined :: forall a. Undefined a -> Boolean
+1. (Easy) Write a new wrapper for the previous JavaScript function with the signature `valuesOfMapGeneric :: forall k v. Map k v -> Either String (Set v)` so it works with a wider variety of maps. Note that you'll need to add some type class constraints for `k` and `v`. The compiler will guide you.
+1. (Medium) Rewrite the earlier `quadraticRoots` function as `quadraticRootsSet` which returns the `Complex` roots as a `Set` via JSON (instead of as a `Pair`).
+1. (Difficult) Rewrite the earlier `quadraticRoots` function as `quadraticRootsSafe` which uses JSON to pass the `Pair` of `Complex` roots over FFI. Don't use the `Pair` constructor in JavaScript, but instead, just return the pair in a decoder-compatible format.
+_Hint_: You'll need to write a `DecodeJson` instance for `Pair`. Consult the [argonaut docs](https://github.com/purescript-contrib/purescript-argonaut-codecs#writing-new-instances) for instruction on writing your own decode instance. Their [decodeJsonTuple](https://github.com/purescript-contrib/purescript-argonaut-codecs/blob/master/src/Data/Argonaut/Decode/Class.purs) instance may also be a helpful reference.  Note that you'll need a `newtype` wrapper for `Pair` to avoid creating an "orphan instance".
+1. (Medium) Write a `decodeArray2D :: String -> Either String (Array (Array Int))` function to parse and decode a JSON string containing a 2D array, such as `"[[1, 2, 3], [4, 5], [6]]"`. _Hint_: You'll need to use `jsonParser` to convert the `String` into `Json` before decoding.
+1. (Medium) The following data type represents a binary tree with values at the leaves:
 
-## Working With Untyped Data
+     ```haskell
+     data Tree a
+       = Leaf a
+       | Branch (Tree a) (Tree a)
+     ```
 
-In this section, we will see how we can use the `Foreign` library to turn untyped data into typed data, with the correct runtime representation for its type.
+     Derive generic `EncodeJson` and `DecodeJson` instances for the `Tree` type.
+     Consult the [argunaut docs](https://github.com/purescript-contrib/purescript-argonaut-codecs#generics) for instructions on how to do this.
+     Note that you'll also need generic instances of `Show` and `Eq` to enable unit testing for this exercise, but those should be straightforward to implement after tackling the JSON instances.
+1. (Difficult) The following `data` type should be represented directly in JSON as either an integer or a string:
 
-The code for this chapter demonstrates how a record can be serialized to JSON and stored in / retrieved from local storage.
+     ```haskell
+     data IntOrString
+       = IntOrString_Int Int
+       | IntOrString_String String
+     ```
 
-The `Main` module defines a type for the saved form data:
+     Write instances of `EncodeJson` and `DecodeJson` for the `IntOrString` data type which implement this behavior. _Hint_: The `alt` operator from `Control.Alt` may be helpful.
+
+## Address book
+
```

Comment:
Should this address book project start with all chapter 8 exercises completed? I think that might be confusing if readers completed those exercises themselves a different way. And those exercises aren't prerequisites for this chapter.